portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,72 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.datastore.Query.*;
+import com.google.gson.Gson;
 import java.io.IOException;
+import javax.servlet.http.*;
 import javax.servlet.annotation.WebServlet;
-import javax.servlet.http.HttpServlet;
-import javax.servlet.http.HttpServletRequest;
-import javax.servlet.http.HttpServletResponse;
+import java.util.List;
+import java.util.ArrayList;
+
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet(urlPatterns = "/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    int maxComments = Integer.parseInt(request.getParameter("max-comments"));
+    String movie = request.getParameter("movie");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    // Filter movieFilter = new FilterPredicate("movie", FilterOperator.EQUAL, movie);
+    // Query query = new Query("Comment").setFilter(movieFilter).addSort("timestamp", SortDirection.DESCENDING);
+    Query query = new Query(movie).addSort("timestamp", SortDirection.DESCENDING);
+    PreparedQuery pq = datastore.prepare(query);
+    List<Entity> resultsList = pq.asList(FetchOptions.Builder.withLimit(maxComments));
+
+    ArrayList<String> comments = new ArrayList<>();
```

Comment:
Try not to use mutable data structures if possible. Typically we try to use ImmutableList.Builder [1] [2].
Same throughout all of the files. 


[1]https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableList.html
[2] https://guava.dev/releases/23.0/api/docs/com/google/common/collect/ImmutableList.Builder.html

Reply:
In this case, since the asList() method returns a List of Entity, do I create an immutable list from this list in the line after, and then use this immutable list throughout the code? Also, for the Arraylist at line 44, right now I'm having to add to it because I want to put all the text properties in one list, would using an Arraylist here be okay?